Sources/SwiftFusion/Inference/Factor.swift

```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+
+/// The most general factor protocol.
+public protocol Factor {
+  var keys: Array<Int> { get set }
+}
+
+/// A `LinearFactor` corresponds to the `GaussianFactor` in GTSAM.
+///
+/// Input is a dictionary of `Key` to `Tensor` pairs, and the output is the scalar
+/// error value. Note here that the Tensor shapes are not checked.
+///
+/// Interpretation
+/// ================
+/// `Input`: the input values as key-value pairs
+///
+public protocol LinearFactor: Factor {
+  typealias ScalarType = Double
+  
+  /// TODO: `Dictionary` still does not conform to `Differentiable`
+  /// Tracking issue: https://bugs.swift.org/browse/TF-899
```

Comment:
https://bugs.swift.org/browse/TF-899

Reply:
What do you think about using an `Array` instead of a `Dictionary`? It's faster and simpler, and it is already differentiable.

You will need a sparse `TangentVector` for `Array`, so that the derivatives don't initialize massive arrays full of zeros. I could add a sparse `TangentVector` to this repo, and I might eventually migrate it to the Swfit stdlib, since it seems generally useful.

(In the future, there may be an option that's even faster than sparse `TangentVector`s: Mutating derivatives that update an `inout TangentVector` in place.)

Reply:
The issue is that every factor only indexes a few variables in the dictionary, and the keys indicate which. I don't think it can be an Array as variables can be vectors of different sizes - depending on what manifold they are TangentVectors of...

Reply:
I think I can work around this by implementing it with a `Dict<Key, Int>` and an `Array<Tensor>` instead, but I am not sure.

Reply:
I'm still not seeing why `Array` is insufficient. Some responses inline.

> The issue is that every factor only indexes a few variables in the dictionary, and the keys indicate which.

With `Array`, factors can index into a small number of variables, as long as we constrain the keys to be consecutive integers starting at `0`. Perhaps the issue is that you often have setups with natural keys that are non-integer or non-consecutive? In that case, I agree that a `Dictionary` makes the most sense.

> I don't think it can be an Array as variables can be vectors of different sizes - depending on what manifold they are TangentVectors of.

The `Element` type of the `Array` can be something general enough to hold all the necessary types. This could be `Tensor` (which holds values of arbitrary dimension) or it could even be `AnyTangentVector` (which holds `TangentVector`s of arbitrary type). `Dictionary` doesn't give you and additional flexibility because it also has a static `Element` type.

> I think I can work around this by implementing it with a `Dict<Key, Int>` and an `Array<Tensor>` instead, but I am not sure.

Yes, this should work. It's a bit complicated though. If you really need `Dictionary` (because of non-integer/non-consecutive keys, or because of some other reason that I haven't understood yet), then I could make a patch pretty quickly that makes `Dictionary` differentiable. Let me know if you think that would be useful!

Reply:
Yes, in GTSAM we have the concept of a Key, which is just an integer, but in many applications we felt the need to have more meaningful keys. We introduced the “Symbol” class to deal with it, although it is a bit of a hack: it concatenates a character and integer into a 64-bit integer. I value your perspective about what the Swift way is. But, indeed, consecutive integers is pretty unnatural.